pkg/socketlb/socketlb.go

```diff
@@ -0,0 +1,171 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package socketlb
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cilium/ebpf"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/cgroups"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/sysctl"
+)
+
+const (
+	Subsystem = "socketlb"
+
+	Connect4     = "cil_sock4_connect"
+	SendMsg4     = "cil_sock4_sendmsg"
+	RecvMsg4     = "cil_sock4_recvmsg"
+	GetPeerName4 = "cil_sock4_getpeername"
+	PostBind4    = "cil_sock4_post_bind"
+	PreBind4     = "cil_sock4_pre_bind"
+	Connect6     = "cil_sock6_connect"
+	SendMsg6     = "cil_sock6_sendmsg"
+	RecvMsg6     = "cil_sock6_recvmsg"
+	GetPeerName6 = "cil_sock6_getpeername"
+	PostBind6    = "cil_sock6_post_bind"
+	PreBind6     = "cil_sock6_pre_bind"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, Subsystem)
+
+	cgroupProgs = []string{
+		Connect4, SendMsg4, RecvMsg4, GetPeerName4,
+		PostBind4, PreBind4, Connect6, SendMsg6,
+		RecvMsg6, GetPeerName6, PostBind6, PreBind6}
+)
+
+// TODO: Clean up bpffs root logic and make this a var.
+func cgroupLinkPath() string {
+	return filepath.Join(bpf.CiliumPath(), Subsystem, "links/cgroup")
+}
+
+// Enable attaches necessary bpf programs for socketlb based on ciliums config.
+//
+// On restart, Enable can also detach unnecessary programs if specific configuration
+// options have changed.
+// It expects bpf_sock.c to be compiled previously, so that bpf_sock.o is present
+// in the Runtime dir.
+func Enable() (err error) {
+	if err := os.MkdirAll(cgroupLinkPath(), 0777); err != nil {
+		return fmt.Errorf("create bpffs link directory: %w", err)
+	}
+
+	spec, err := bpf.LoadCollectionSpec(filepath.Join(option.Config.StateDir, "bpf_sock.o"))
+	if err != nil {
+		return fmt.Errorf("failed to load collection spec for bpf_sock.o: %w", err)
+	}
+
+	if err := bpf.StartBPFFSMigration(bpf.TCGlobalsPath(), spec); err != nil {
+		return fmt.Errorf("failed to start bpffs map migration: %w", err)
+	}
+
+	// This captures named return variable err.
+	defer func() {
+		if err != nil {
+			log.WithError(err).Debug("Reverting bpffs map migration")
+			if e := bpf.FinalizeBPFFSMigration(bpf.TCGlobalsPath(), spec, true); e != nil {
+				log.WithError(e).Error("Could not revert bpffs map migration")
+				return
+			}
+		}
+
+		log.Debug("Finalizing bpffs map migration")
+		if e := bpf.FinalizeBPFFSMigration(bpf.TCGlobalsPath(), spec, false); e != nil {
+			log.WithError(e).Error("Could not finalize bpffs map migration")
+		}
+	}()
+
+	coll, err := bpf.LoadCollection(spec, ebpf.CollectionOptions{
+		Maps: ebpf.MapOptions{PinPath: bpf.TCGlobalsPath()},
+	})
+	var ve *ebpf.VerifierError
+	if errors.As(err, &ve) {
+		log.Debugf("verifier error: %+v", ve)
```

Comment:
^ I see that this is at Warning level now, but I suspect it will be unreadable since the buffer has a bunch of `\n` in it, and these are typically not rendered by default into the log. The result is a wall of text on one line which is really difficult to deal with.

I'm not sure what we're doing today, but existing verifier errors show up with each line from the kernel as a separate line in the log. This makes it easy to read when something goes wrong. It would be nice to make sure that these logs print out in a readable manner as well (relevant also to https://github.com/cilium/cilium/pull/24506)